File: examples/sequential-java/Example1.java
/* This code adds all the values of an array */

public class Example1 {
	private static final long NUM_RECTS = 1_000_000_000;
	private double result;
	
	public Example1() {
		this.result = 0;
	}
	
	public double getResult() {
		return result;
	}
	
	public void calculate() {
		double sum, width, mid, height;
		
		sum = 0;
		width = 1.0 / (double) NUM_RECTS;
		for (int j = 0; j < NUM_RECTS; j++) {
			mid = (j + 0.5) * width;
			height = 4.0 / (1.0 + (mid * mid));
			sum += height;
		}
		result = width * sum;
	}
	
	public static void main(String args[]) {
		long startTime, stopTime;
		double acum = 0;
		
		Example1 e = new Example1();
		acum = 0;
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();
			e.calculate();
			stopTime = System.currentTimeMillis();
			acum += (stopTime - startTime);
		}
		System.out.printf("PI = %d\n", e.getResult());
		System.out.printf("avg time = %.5f\n", (acum / Utils.N));
	}
}
			

File: examples/tbb/example1.cpp
/* This code calculates an approximation of PI */

#include <iostream>
#include "utils/cppheader.h"
#include <tbb/task_scheduler_init.h>
#include <tbb/parallel_reduce.h>
#include <tbb/blocked_range.h>

using namespace std;
using namespace tbb;

const long NUM_RECTS = 100000000; //1e8
const int GRAIN = 100000; //1e6

class CalculatingPi {
public:
	double sum;

public:
	CalculatingPi() : sum(0) {}

	CalculatingPi(CalculatingPi &x, split) 
		: sum(0) {} 

	void operator() (const blocked_range<int> &r) {
		double mid, height, width;
		
		sum = 0;
		width = 1.0 / (double) NUM_RECTS;
		for (int i = r.begin(); i != r.end(); i++) {
			mid = (i + 0.5) * width;
			height = 4.0 / (1.0 + (mid * mid));
			sum += height;
		}
	}

	void join(const CalculatingPi &x) {
		sum += x.sum;
	}
};

int main(int argc, char* argv[]) {
	Timer t;
	double ms, area;

	cout << "Starting..." << endl;
	ms = 0;
	for (int i = 0; i < N; i++) {
		t.start();

		CalculatingPi cp;
		parallel_reduce(blocked_range<int>(0, NUM_RECTS, GRAIN), cp);
		area = cp.sum * (1.0 / (double) NUM_RECTS);
		ms += t.stop();
	}
	cout << "PI = " << area << endl;
	cout << "avg time = " << (ms / N) << " ms" << endl;

	return 0;
}
